File: leetcode/216-combination-sum-iii/main.go
package main

import "fmt"

// 1st approach
// classic backtracking problem
// very similar to euler/31
// Time		O(nCk) pick K numbers from N numbers
// Space	O(nCk) the recursive call stack
// beats 100%
func combinationSum3(k int, n int) [][]int {
	result := [][]int{}
	var dfs func(num int, path []int, fromNum int)
	dfs = func(num int, path []int, fromNum int) {
		if num == 0 && len(path) == k {
			result = append(result, path)
		} else if len(path) < k {
			for i := fromNum; i < 10; i++ {
				nextPath := []int{}
				nextPath = append(nextPath, path...)
				nextPath = append(nextPath, i)
				dfs(num-i, nextPath, i+1)
			}
		}
	}
	dfs(n, []int{}, 1)
	return result
}

func main() {
	fmt.Println(combinationSum3(3, 7))
	fmt.Println(combinationSum3(3, 9))
	fmt.Println(combinationSum3(5, 19))
}

File: leetcode/216-combination-sum-iii/main.py
"""
    1st: recursion
    - similar to lc77(combinations)

    Time    O(nCk)
    Space   O(nCk)
    80 ms, faster than 32.73% 
"""


class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        nums = [i+1 for i in range(9)]
        self.res = []
        self.dfs([], nums, k, n)
        return self.res

    def dfs(self, chosen, nums, k, n):
        if len(chosen) == k:
            if n == 0:
                self.res.append(chosen)
            return
        for i in range(len(nums)):
            x = nums[i]
            self.dfs(chosen + [x], nums[i+1:], k, n-x)
